forte/data/data_store.py

```diff
@@ -1226,18 +1258,29 @@ def co_iterator_annotation_like(
         # Initialize the first entry of all entry lists
         # it avoids empty entry lists or non-existent entry list
         first_entries = []
+        all_entries = {}
 
-        for tn in type_names:
+        if range_annotation:
+            for entry_type in type_names:
+                all_entries[entry_type] = get_bisect_range(
+                    self.__elements[entry_type]
+                )
```

Comment:
I think this is inefficient here: `co_iterator` is designed to be efficient,  it has an `O(KN)` complexity, where K is the number of types, N is the length of the list.  Here, when we do `get_bisect_range` we don't have to actually do the list slicing operation `search_list[begin_index:end_index]`, we simply need to return `begin_index` and `end_index`, and then do the iteration within this range, to avoid this extra slicing.

Reply:
Based on @hepengfe's [comments](https://github.com/asyml/forte/pull/849#discussion_r919548111), it looks like we cannot just use `begin_index` and `end_index` to specify the range?

Reply:
My understanding of that comment is that we need to use both of them together? What else is needed to specify the range?

Reply:
[Previously](https://github.com/mylibrar/forte/blob/0ca9602d3d287beb2521584f5fc50c2f5905cebc/forte/data/data_pack.py#L1087) we used `[range_begin, range_begin] <= [anno.begin, anno.end] <= [range_end, range_end]` to narrow down the search, but it can be problematic since we might include unexpected annotations with `range_begin<=anno.begin<range_end and range_end<anno.end`. [Here](https://github.com/asyml/forte/pull/849/files#diff-f1c68405498aa1cba27f81862f28d3edecfdc2d8501394bb3cac9d83dfa1283dR1217) is the latest implementation.

Reply:
@Pushkar-Bhuse  please make sure the implementation here is correct according to the comments @mylibrar  and @hepengfe. The unexpected change of length in the test case may be caused by the implementation.

Reply:
Yes I have addressed all the changes and fixed the test cases.